app/web_framework/cluster_application.py

```diff
@@ -15,6 +52,33 @@ def get_all_handlers(root_route, default_params):
         :rtype: list [tuple (str, tornado.web.RequestHandler, dict)]
         """
         all_route_nodes = [root_route] + root_route.descendants()
+
+        # Map regex to all the routes that have the same regex pattern.
+        route_map = {}
+        for route in all_route_nodes:
+            route_path = route.regex()
+            if route_path not in route_map:
+                route_map[route_path] = []
+            route_map[route_path].append(route)
+
+        # Map regex to a list of route and RequestHandlerWrapper tuples.
+        # The RequestHandlerWrapper knows all the RequstHandlers and what API version those handlers work
+        # with for a given route, so when a route is hit by the API, we know which RequestHandler to use.
+        route_handler_map = {}
+        for regex in route_map:
+            for route in route_map[regex]:
+                if regex not in route_handler_map:
+                    route_handler_map[regex] = []
+                # Create a dict that maps version to handler
+                handlers_map = dict([(route.version, route.handler) for route in route_map[regex]])
```

Comment:
I have been staring at this for awhile so I might be looking at this wrong, but I think there might be a bug here. This dict is created inside this `for route in route_map[regex]` loop, but it is doing the same loop as a comprehension. It seems like this would create copies of the same handlers_map. Am I seeing that right?

I have a feeling the logic in this method can be simplified. Maybe the behavior would be more obvious with some test cases.